src/app/_hooks/useSort.ts

```diff
@@ -1,71 +1,49 @@
 import { useMemo } from 'react'
 
 import { type NextServerSearchParams } from '@/types/searchParams'
-import type {
-  AlphabeticalSortId,
-  ChronologicalSortId,
-  SortableByDate,
-  SortId,
-} from '@/types/sortTypes'
+import type { SortConfig } from '@/types/sortTypes'
 import { type Object } from '@/types/utils'
 
 import { SORT_KEY } from '@/constants/searchParams'
-import {
-  ALPHABETICAL_SORT_IDS,
-  CHRONOLOGICAL_SORT_IDS,
-  VALID_SORT_IDS,
-} from '@/constants/sortConstants'
+import { VALID_SORT_KEYS } from '@/constants/sortConstants'
 
 import { normalizeQueryParam } from '@/utils/queryUtils'
-import { sortEntriesAlphabetically } from '@/utils/sortEntriesAlphabetically'
-import { sortEntriesByDate } from '@/utils/sortEntriesByDate'
 
 type UseSortProps<Entry extends Object> = {
   searchParams: NextServerSearchParams
   entries: Array<Entry>
-  sortBy: keyof SortableByDate & keyof Entry
-  defaultSortId: SortId
+  configs: ReadonlyArray<SortConfig<Entry>>
 }
 
 export function useSort<Entry extends Object>({
   searchParams,
   entries,
-  sortBy,
-  defaultSortId,
+  configs,
 }: UseSortProps<Entry>) {
-  const normalizedQuery = normalizeQueryParam(searchParams, SORT_KEY)
-  const validatedSortId = getValidatedSortId(normalizedQuery, defaultSortId)
+  const defaultConfig = configs[0]
 
-  const sortedResults = useMemo(() => {
-    if (ALPHABETICAL_SORT_IDS.includes(validatedSortId)) {
-      return sortEntriesAlphabetically({
-        entries,
-        sortId: validatedSortId as AlphabeticalSortId,
-      })
-    }
+  if (!defaultConfig) {
+    throw new Error('At least one sort config must be provided')
+  }
```

Comment:
I feel like at this point in the code - it shouldn't be possible for `defaultConfig` to be `undefined`? Can't TS magic help us here?

```
type UseSortProps<Entry extends Object> = {
  searchParams: NextServerSearchParams
  entries: Array<Entry>
  configs: readonly [SortConfig<Entry>, ...Array<SortConfig<Entry>>]
}
```

```
export const blogSortData: readonly [
  SortConfig<BlogPost>,
  ...Array<SortConfig<BlogPost>>,
] = [
  {
    key: 'newest',
    label: 'Newest',
    sortFn: sortPostsByDateDesc,
  },
  {
    key: 'oldest',
    label: 'Oldest',
    sortFn: sortPostsByDateAsc,
  },
] as const
```

Reply:
Yes absolutely! For some reason, I couldn't make it work. Let me try that again :)